core/src/Streamly/Internal/Data/MutArray.hs

```diff
@@ -83,10 +87,100 @@ getSlicesFromLen :: forall m a. (Monad m, Unbox a)
-{-# INLINE compactChunksLE #-}
-compactChunksLE :: (MonadIO m, Unbox a) =>
+{-# INLINE compactLE #-}
+compactLE :: (MonadIO m, Unbox a) =>
     Int -> Stream m (MutArray a) -> Stream m (MutArray a)
--- compactLE n = Stream.catRights . Stream.parseManyD (pCompactChunksLE n)
-compactChunksLE = rCompactChunksLE
+-- XXX compactLE can be moved to MutArray/Type if we are not using the parser
+-- to implement it.
+-- compactLE n = Stream.catRights . Stream.parseManyD (pCompactLE n)
+compactLE = rCompactLE
+
+data SplitState s arr
+    = Initial s
+    | Buffering s arr
+    | Splitting s arr
+    | Yielding arr (SplitState s arr)
+    | Finishing
+
+-- | Split a stream of arrays on a given separator byte, dropping the separator
+-- and coalescing all the arrays between two separators into a single array.
+--
+{-# INLINE_NORMAL _compactOnByteCustom #-}
+_compactOnByteCustom
+    :: MonadIO m
+    => Word8
+    -> Stream m (MutArray Word8)
+    -> Stream m (MutArray Word8)
+_compactOnByteCustom byte (Stream.Stream step state) =
+    Stream.Stream step' (Initial state)
+
+    where
+
+    {-# INLINE_LATE step' #-}
+    step' gst (Initial st) = do
+        r <- step gst st
+        case r of
+            Stream.Yield arr s -> do
+                (arr1, marr2) <- breakOn byte arr
+                return $ case marr2 of
+                    Nothing   -> Stream.Skip (Buffering s arr1)
+                    Just arr2 -> Stream.Skip (Yielding arr1 (Splitting s arr2))
+            Stream.Skip s -> return $ Stream.Skip (Initial s)
+            Stream.Stop -> return Stream.Stop
+
+    step' gst (Buffering st buf) = do
+        r <- step gst st
+        case r of
+            Stream.Yield arr s -> do
+                (arr1, marr2) <- breakOn byte arr
+                -- XXX Use spliceExp instead and then rightSize?
+                buf1 <- splice buf arr1
+                return $ case marr2 of
+                    Nothing -> Stream.Skip (Buffering s buf1)
+                    Just x -> Stream.Skip (Yielding buf1 (Splitting s x))
+            Stream.Skip s -> return $ Stream.Skip (Buffering s buf)
+            Stream.Stop -> return $
+                if byteLength buf == 0
+                then Stream.Stop
+                else Stream.Skip (Yielding buf Finishing)
+
+    step' _ (Splitting st buf) = do
+        (arr1, marr2) <- breakOn byte buf
+        return $ case marr2 of
+                Nothing -> Stream.Skip $ Buffering st arr1
+                Just arr2 -> Stream.Skip $ Yielding arr1 (Splitting st arr2)
+
+    step' _ (Yielding arr next) = return $ Stream.Yield arr next
+    step' _ Finishing = return Stream.Stop
+
+-- XXX implement predicate based version of this
+-- XXX Naming of predicate based vs custom version
+
+-- | Split a stream of arrays on a given separator byte, dropping the separator
+-- and coalescing all the arrays between two separators into a single array.
+--
+{-# INLINE compactOnByte #-}
+compactOnByte
```

Comment:
Is `compact` the right term for this? Maybe `words`?

Reply:
Does `compactBetweenByte` sound better?

Reply:
compact is a compact word for compacting multiple arrays into a bigger one.